internal/concierge/impersonator/impersonator.go

```diff
@@ -0,0 +1,153 @@
+	return newInternal(cache, log, rest.InClusterConfig)
+}
+
+func newInternal(cache *authncache.Cache, log logr.Logger, getConfig func() (*rest.Config, error)) (*Proxy, error) {
+	kubeconfig, err := getConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not get in-cluster config: %w", err)
+	}
+
+	serverURL, err := url.Parse(kubeconfig.Host)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse host URL from in-cluster config: %w", err)
+	}
+
+	kubeTransportConfig, err := kubeconfig.TransportConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not get in-cluster transport config: %w", err)
+	}
+	kubeTransportConfig.TLS.NextProtos = []string{"http/1.1"}
+
+	kubeRoundTripper, err := transport.New(kubeTransportConfig)
+	if err != nil {
+		return nil, fmt.Errorf("could not get in-cluster transport: %w", err)
+	}
+
+	proxy := httputil.NewSingleHostReverseProxy(serverURL)
+	proxy.Transport = kubeRoundTripper
+
+	return &Proxy{
+		cache: cache,
+		proxy: proxy,
+		log:   log,
+	}, nil
+}
+
+func (p *Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	log := p.log.WithValues(
+		"url", r.URL.String(),
+		"method", r.Method,
+	)
+
+	tokenCredentialReq, err := extractToken(r)
+	if err != nil {
+		log.Error(err, "invalid token encoding")
+		http.Error(w, "invalid token encoding", http.StatusBadRequest)
+		return
+	}
+	log = log.WithValues(
+		"authenticator", tokenCredentialReq.Spec.Authenticator,
+		"authenticatorNamespace", tokenCredentialReq.Namespace,
+	)
+
+	userInfo, err := p.cache.AuthenticateTokenCredentialRequest(r.Context(), tokenCredentialReq)
+	if err != nil {
+		log.Error(err, "received invalid token")
+		http.Error(w, "invalid token", http.StatusUnauthorized)
+		return
+	}
+
+	if userInfo == nil {
+		log.Info("received token that did not authenticate")
+		http.Error(w, "not authenticated", http.StatusUnauthorized)
+		return
+	}
+	log = log.WithValues("userID", userInfo.GetUID())
+
+	newHeaders := getProxyHeaders(userInfo, r.Header)
+	r.Header = newHeaders
+
+	log.Info("proxying authenticated request")
+	p.proxy.ServeHTTP(w, r)
+}
+
+func getProxyHeaders(userInfo user.Info, requestHeaders http.Header) http.Header {
+	newHeaders := http.Header{}
+	newHeaders.Set("Impersonate-User", userInfo.GetName())
+	for _, group := range userInfo.GetGroups() {
+		newHeaders.Add("Impersonate-Group", group)
+	}
+	for _, header := range allowedHeaders {
+		values := requestHeaders.Values(header)
+		for i := range values {
+			newHeaders.Add(header, values[i])
+		}
+	}
+	return newHeaders
+}
+
+func extractToken(req *http.Request) (*login.TokenCredentialRequest, error) {
+	authHeader := req.Header.Get("Authorization")
+	if authHeader == "" {
+		return nil, fmt.Errorf("missing authorization header")
+	}
+	if !strings.HasPrefix(authHeader, "Bearer ") {
+		return nil, fmt.Errorf("authorization header must be of type Bearer")
+	}
+	encoded := strings.TrimPrefix(authHeader, "Bearer ")
```

Comment:
The kube authentiactor code already does this for you.

Reply:
This came out a little messier than I expected, so I pushed the commit to a separate branch for you to take a look at before I pushed this to the development branch: https://github.com/ankeesler/pinniped/commit/04af5b767db2d834657a6765477faf1281c77695. What do you think of just copying those `Authorization` header parsing lines from the `bearertoken.Authenticator.AuthenticateRequest()` function? Here are some of the tradeoffs:

Pros of using `bearertoken.Authenticator` in our code:
- We know it works
- We know it will be patched if there is a header parsing issue

Cons of using `bearertoken.Authenticator` in our code:
- We lose some very specific error messages about parsing the `Authorization` header in our logs
- It is harder to communicate response message/code back to the caller because of generic return errors
- Using the `logr.Logger` pattern is a little more confusing to follow (shouldn't we be using `plog` everywhere though?)

Reply:
Mmm. Ignore me, I think. Because of the conversation we had about using `"k8s.io/apiserver/pkg/authentication/token/cache"`, I think we will want this `bearertoken.Authenticator` here.

Reply:
b3cdc438ce0511eeb8e39aa93a0fdf940c0b9d7d